db/migrate/20230126205325_create_roles.rb

```diff
@@ -0,0 +1,18 @@
+class CreateRoles < ActiveRecord::Migration[7.0]
+  def change
+    create_table(:roles) do |t|
+      t.string :name
+      t.references :resource, polymorphic: true
+
+      t.timestamps
+    end
+
+    create_table(:users_roles, id: false) do |t|
+      t.references :user
+      t.references :role
+    end
+
+    add_index(:roles, [ :name, :resource_type, :resource_id ])
+    add_index(:users_roles, [ :user_id, :role_id ])
```

Comment:
```suggestion
    create_table :roles  do |t|
      t.string :name, null: false
      t.references :resource, index: true, polymorphic: true
      t.timestamps
      t.index %i[name resource_id resource_type], unique: true
    end

    create_table :user_roles, id: false do |t|
      t.references :user, foreign_key: true, index: true, null: false
      t.references :role, foreign_key: true, index: true, null: false
      t.index %i[user_id role_id], unique: true
    end
```

Some migration cleanup/making the columns tighter (non-null constraints, unique indices and the like)